guvnor-ala/guvnor-ala-registry-local/src/main/java/org/guvnor/ala/registry/local/InMemoryRuntimeRegistry.java

```diff
@@ -0,0 +1,171 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.guvnor.ala.registry.local;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.guvnor.ala.registry.RuntimeRegistry;
+import org.guvnor.ala.runtime.Runtime;
+import org.guvnor.ala.runtime.RuntimeId;
+import org.guvnor.ala.runtime.providers.Provider;
+import org.guvnor.ala.runtime.providers.ProviderId;
+import org.guvnor.ala.runtime.providers.ProviderType;
+
+import static org.uberfire.commons.validation.PortablePreconditions.*;
+
+/**
+ * @TODO: This is a not thread-safe implementation for local testing. A
+ * more robust and distributed implementation should be provided for real
+ * use cases. All the lookups mechanisms and structures needs to be improved for performance.
+ */
+@ApplicationScoped
+public class InMemoryRuntimeRegistry implements RuntimeRegistry {
+
+    private final Map<String, ProviderType> providerTypes;
+    private final Map<String, Provider> providers;
+    private final Map<ProviderType, List<Provider>> providersByType;
+    private final Map<ProviderType, List<Runtime>> runtimesByProviderType;
+
+    public InMemoryRuntimeRegistry() {
+        providerTypes = new HashMap<>();
+        providers = new HashMap<>();
+        providersByType = new HashMap<>();
+        runtimesByProviderType = new HashMap<>();
+    }
+
+    @Override
+    public void registerProviderType( ProviderType pt ) {
+        providerTypes.put( pt.getProviderTypeName(), pt );
+    }
+
+    @Override
+    public List<ProviderType> getAllProviderTypes() {
+        return new ArrayList<>( providerTypes.values() );
+    }
+
+    @Override
+    public ProviderType getProviderType( String provider ) {
+        return providerTypes.get( provider );
+    }
+
+    @Override
+    public void unregisterProviderType( ProviderType providerType ) {
+        providerTypes.remove( providerType.getProviderTypeName() );
+    }
+
+    @Override
+    public void registerProvider( Provider provider ) {
+        providers.put( provider.getId(), provider );
+        if ( !providersByType.containsKey( provider.getProviderType() ) ) {
+            providersByType.put( provider.getProviderType(), new ArrayList<>() );
+        }
+        providersByType.get( provider.getProviderType() ).add( provider );
```

Comment:
You can use the `computeIfAbsent` method introduced in Java 8 to avoid the above conditional and instead replace this line with:

    providersByType.computeIfAbsent( provider.getProviderType(), providerType -> new ArrayList<>() ).add( provider );